src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,411 @@
+     */
+    private NBTTagCompound serializedForm;
+
+    private RegionSnapshot(World world, Region region, ImmutableList<IBlockState> blockStates, ImmutableList<BlockSnapshot> tileSnapshots) {
+        this.world = world;
+        this.region = region;
+        this.blockStates = blockStates;
+        this.tileSnapshots = tileSnapshots;
+    }
+
+    public boolean restore() {
+        boolean statesSucceeded = true;
+        int index = 0;
+        for (BlockPos pos : region) {
+            IBlockState state = blockStates.get(index);
+            if (state == DummyBlockState.INSTANCE)
+                continue;
+
+            statesSucceeded &= world.setBlockState(pos, state);
+
+            index++;
+        }
+
+        boolean snapshotsSucceeded = true;
+        for (BlockSnapshot snapshot : tileSnapshots) {
+            snapshotsSucceeded &= snapshot.restore();
+        }
+
+        return statesSucceeded && snapshotsSucceeded;
+    }
+
+    public Region getRegion() {
+        return region;
+    }
+
+    /**
+     * From minimum Z to maximum Z; then from minimum Y to maximum Y; then from minimum X to maximum X; each entry are
+     * stored in the above order and represent the block state in that position. If an entry is {@code null}, it means
+     * that position should not be replaced regularly: it might be replaced with a {@link BlockSnapshot}, or it should
+     * be left untouched.
+     */
+    public ImmutableList<IBlockState> getBlockStates() {
+        return blockStates;
+    }
+
+    /**
+     * The indices of the entries of this list is unrelated to any positions, instead, the entry stores the affected
+     * coordinate.
+     */
+    public ImmutableList<BlockSnapshot> getTileEntities() {
+        return tileSnapshots;
+    }
+
+    public NBTTagCompound serialize() {
+        return serializeTo(new NBTTagCompound());
+    }
+
+    public NBTTagCompound serializeTo(NBTTagCompound tag) {
+        if (serializedForm == null)
+            serializedForm = serialize(tag, this);
+        return serializedForm;
+    }
+
+    private static final class Builder {
+
+        private static final BiPredicate<BlockPos, IBlockState> AIR_CHECK = (pos, state) -> state.getMaterial() != Material.AIR;
+
+        private World world;
+        private Region region;
+        private BiPredicate<BlockPos, IBlockState> normalValidator;
+        private TriPredicate<BlockPos, IBlockState, TileEntity> tileValidator;
+
+        private boolean built = false;
+
+        private Builder(World world, Region region) {
+            this.world = world;
+            this.region = region;
+        }
+
+        /**
+         * Set validator for normal blocks (non tile entities).
+         *
+         * @return this
+         */
+        public Builder checkBlocks(BiPredicate<BlockPos, IBlockState> normalValidator) {
+            Preconditions.checkState(!built);
+
+            this.normalValidator = normalValidator;
```

Comment:
If you ask me it's counterintuitive not to and it with an existing validator...
Additionally anding here would save you quite a few checks, before calling this Method.